File: leetcode/848_Shifting_Letters/Shifting_Letters.java
class Solution {
    public String shiftingLetters(String S, int[] shifts) {
        char[] res = new char[S.length()];
        long shift = 0;
        for (int i = shifts.length - 1; i >= 0; i--) {
            shift += shifts[i];
            long tmp = shift % 26;
            if (tmp > ('z' - S.charAt(i))) {
                tmp -= 'z' - S.charAt(i) + 1;
                res[i] = (char)('a' + tmp);
            } else res[i] = (char)(S.charAt(i) + tmp);
        }
        return new String(res);
    }
}

File: leetcode/848_Shifting_Letters/Shifting_Letters.cpp
class Solution {
public:
    string shiftingLetters(string S, vector<int>& shifts) {
        string res(S);
        int tmp = 0;
        for (int i = shifts.size() - 1; i >= 0; i--) {
            tmp += shifts[i] % 26;
            tmp = tmp % 26;
            if (tmp > 'z' - S[i]) res[i] = 'a' + tmp - 1 - 'z' + S[i];
            else res[i] += tmp;
        }
        return res;
    }
};
